### PR TITLE
Client Automatically Opens Start Menu

### PR DESCRIPTION
**Pull request by AMZN-Gene**
- GameLauncher.exe setreg update to load the start menu level
- Converting existing ScriptCanvas start menu to C++
- Adding start up messaging when attempting to connect to host and when connection fails
- Disabling the host button for launchers that don't support hosting

### CODE REVIEW

**Review by kberg-amzn — APPROVED** [review]
Just out of curiosity, not asking to make the change at this point, but is the native component actually necessary? What is missing from our script-bind that prevents all of this from being implemented in script canvas?
Otherwise, awesome! Much needed changes for release.

File: Registry/settings.multiplayersample_gamelauncher.setreg [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+    "Amazon": {
+        "AzCore": {
+            "Runtime": {
+                "ConsoleCommands": {
+                    "loadlevel": "startmenu"
```
Comment (AMZN-Olex) [inline Registry/settings.multiplayersample_gamelauncher.setreg:6]: Does this apply to Server Launcher?

Comment (AMZN-Gene) [reply Registry/settings.multiplayersample_gamelauncher.setreg:6]: Nay, this PR only adds to __gamelauncher.setreg_

Suppose we could also add this to server launchers... I'd want to make sure loading a level in setreg for servers doesn't override editor-server does when it launches its server. Last I checked, setting the server setreg loadlevel _would_ load that lever, and then switch over to the original level the editor was attempting to test. But ideally we'd want to a way to suppress that setreg loadlevel when the editor runs CTRL+G.

**Comment by AMZN-Gene** [comment]
> Just out of curiosity, not asking to make the change at this point, but is the native component actually necessary? What is missing from our script-bind that prevents all of this from being implemented in script canvas? Otherwise, awesome! Much needed changes for release.

Probably doable in script canvas, but I'd avoid writing anything one-off fire-and-forget type responses to events in script canvas. Large graphs aren't easy on the eyes, and sometimes graph assets break and are difficult to debug, like recently seen with reticle graph breaking during asset load inside lua.

**Review by AMZN-stankowi — APPROVED** [review]

**Review by AMZN-alexpete — APPROVED** [review]

File: Gem/Code/Source/Components/UI/UiStartMenuComponent.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        }
+
+        if (buttonEntityId == m_exitButtonUi)
+        {
+            console->PerformCommand("quit");
+        }
+        else if (buttonEntityId == m_hostButtonUi)
+        {
+            console->PerformCommand("host");
+            console->PerformCommand("loadlevel NewStarbase");
```
Comment (mbalfour-amzn) [inline Gem/Code/Source/Components/UI/UiStartMenuComponent.cpp:102]: Really feels like the next level to load should at least be a parameter on the component.

Comment (AMZN-Gene) [reply Gem/Code/Source/Components/UI/UiStartMenuComponent.cpp:102]: Originally had that as a variable in the UI, but felt like a variable I wouldn't want art to accidentally change/misspell. Maybe a class static constexpr would be more appropriate.

File: Gem/Code/Source/Components/UI/UiStartMenuComponent.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+#pragma once
+
+#include <AzCore/Component/Component.h>
+#include <AzCore/Settings/SettingsRegistry.h>
+#include <Multiplayer/IMultiplayer.h>
+
+namespace MultiplayerSample
+{
+    class UiStartMenuComponent
```
Comment (lmbr-pip) [inline Gem/Code/Source/Components/UI/UiStartMenuComponent.h:17]: nit: comment on all public methods please

**Review by lmbr-pip — COMMENTED** [review]
From a UX POV, do we autofill the server and port if they are set by cvars?

**Comment by AMZN-Gene** [comment]
> From a UX POV, do we autofill the server and port if they are set by cvars?

I would LOVE to! I couldn't seem to access a valid cvar on activate from this component. I _could_ add a few ticks and probably grab it, but I really need to debug why the cvars aren't available. The UI is activated well after all systems components are stood up.

File: Gem/Code/Source/Components/UI/UiStartMenuComponent.cpp (hunk 2 above)
Comment (AMZN-Gene) [reply Gem/Code/Source/Components/UI/UiStartMenuComponent.cpp]: fixed!

Comment (mbalfour-amzn) [reply Gem/Code/Source/Components/UI/UiStartMenuComponent.cpp]: "Originally had that as a variable in the UI, but felt like a variable I wouldn't want art to accidentally change/misspell." - you can make it an Asset Reference that can only be set to a Spawnable type. This solves the misspelling problem and also gives the asset dependency system more information for creating bundles.

**Pull request merged**